core/config/stop_config.cpp

```diff
@@ -0,0 +1,151 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+#include <ginkgo/core/stop/residual_norm.hpp>
+#include <ginkgo/core/stop/time.hpp>
+
+
+#include "core/config/config_helper.hpp"
+#include "core/config/dispatch.hpp"
+
+
+namespace gko {
+namespace config {
+
+
+inline deferred_factory_parameter<stop::CriterionFactory> configure_time(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    auto factory = stop::Time::build();
+    if (auto& obj = config.get("time_limit")) {
+        factory.with_time_limit(gko::config::get_value<long long int>(obj));
+    }
+    return factory;
+}
+
+
+inline deferred_factory_parameter<stop::CriterionFactory> configure_iter(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    auto factory = stop::Iteration::build();
+    if (auto& obj = config.get("max_iters")) {
+        factory.with_max_iters(gko::config::get_value<size_type>(obj));
+    }
+    return factory;
+}
+
+
+stop::mode get_mode(const std::string& str)
+{
+    if (str == "absolute") {
+        return stop::mode::absolute;
+    } else if (str == "initial_resnorm") {
+        return stop::mode::initial_resnorm;
+    } else if (str == "rhs_norm") {
+        return stop::mode::rhs_norm;
+    }
+    GKO_INVALID_STATE("Not valid " + str);
+}
+
+
+template <typename ValueType>
+class ResidualNormConfigurer {
+public:
+    static deferred_factory_parameter<
+        typename stop::ResidualNorm<ValueType>::Factory>
```

Comment:
Is the concrete type necessary, wouldn't `CriterionFactory` be enough?

Reply:
Yes, the CriterionFactory should be also enough but I prefer the concrete type when it does not require more stuff.